File: robots/robot.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>

// Function to recursively list directories and files with full paths
void listDirectory(const char *basePath, int level, FILE *file) {
    struct dirent *entry;
    DIR *dir = opendir(basePath);

    if (dir == NULL) {
        perror("Unable to open directory");
        return;
    }

    // Loop through all entries in the directory
    while ((entry = readdir(dir)) != NULL) {
        // Skip the "." and ".." entries
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0) {
            continue;
        }

        // Build the full path for each entry
        char path[1024];
        snprintf(path, sizeof(path), "%s/%s", basePath, entry->d_name);

        // Print the full path with indentation for hierarchy
        for (int i = 0; i < level; i++) {
            fprintf(file, "");  // Indentation for visual hierarchy
        }
        fprintf(file, "%s\n", path);  // Print the full path of the directory or file

        // Check if the entry is a directory
        struct stat statbuf;
        if (stat(path, &statbuf) == 0 && S_ISDIR(statbuf.st_mode)) {
            // Recursively list the subdirectory
            listDirectory(path, level + 1, file);
        }
    }

    closedir(dir);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <directory_path>\n", argv[0]);
        return 1;
    }

    // Open a text file to write the directory structure
    FILE *file = fopen("directory_structure.txt", "w");
    if (file == NULL) {
        perror("Unable to create file");
        return 1;
    }

    // Start listing from the specified path
    listDirectory(argv[1], 0, file);

    // Close the file
    fclose(file);

    printf("Directory structure saved in directory_structure.txt\n");
    return 0;
}

File: robots/nodejs/robot.js
const fs = require('fs');
const path = require('path');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');
const os = require('os');
const numThreads = os.cpus().length;  // Number of CPU cores to spawn worker threads

// Recursive function to list directories and files
function listDirectory(dirPath, level, fileStream) {
    fs.readdir(dirPath, { withFileTypes: true }, (err, files) => {
        if (err) {
            console.error(`Error reading directory: ${dirPath}`);
            return;
        }

        files.forEach(file => {
            // Full path of the file or directory
            const fullPath = path.join(dirPath, file.name);

            // Indentation based on depth (level)
            //const indentation = ''.repeat(level * 4);
			const indentation = '';
            fileStream.write(`${indentation}${fullPath}\n`);

            // If it's a directory, process it recursively
            if (file.isDirectory()) {
                // Spawn worker threads for parallel processing
                if (isMainThread) {
                    const worker = new Worker(__filename, {
                        workerData: { dirPath: fullPath, level: level + 1 }
                    });

                    worker.on('message', () => {});
                    worker.on('error', (error) => console.error(error));
                    worker.on('exit', (code) => {
                        if (code !== 0) {
                            console.error(`Worker stopped with exit code ${code}`);
                        }
                    });
                } else {
                    listDirectory(workerData.dirPath, workerData.level, fileStream);
                }
            }
        });
    });
}

// Main thread starts here
if (isMainThread) {
    const directoryPath = process.argv[2] || '.';  // Default to current directory
    const outputFilePath = 'directory_structure.txt';

    // Open the output file stream
    const fileStream = fs.createWriteStream(outputFilePath, { flags: 'w' });

    // Start the listing process in the main thread
    console.log('Listing directory structure...');
    listDirectory(directoryPath, 0, fileStream);

    // Close the file stream when done
    fileStream.end(() => {
        console.log(`Directory structure saved in ${outputFilePath}`);
    });
} else {
    // This is for worker threads to run the listDirectory function
    listDirectory(workerData.dirPath, workerData.level, fs.createWriteStream('directory_structure.txt', { flags: 'a' }));
}
